streamr-dev/streamr-platform
# [PLATFORM-758] Don't perform variadic handling on subcanvases

Fixes an issue preventing a production canvas from working. 

See: https://streamr.atlassian.net/browse/PLATFORM-758

Production canvas in question rendering locally with the fixes in this PR:

![image](https://user-images.githubusercontent.com/43438/58380163-6aa34e80-7fe0-11e9-8a5a-c50747d0fc81.png)

To test: 

1. Create a new canvas (canvas 1)
2. Add the PassThrough module
3. Export the input on the PassThrough (bonus: note it adds a new placeholder input)
4. Create another new canvas.
5. Add canvas module.
6. Select canvas 1 as subcanvas. 
7. Note it doesn't break. 

Current `development` breaks at step 7 with same error as reported in JIRA ticket.

Alternatively you could start the app pointed at the production endpoints:

```
STREAMR_API_URL=https://www.streamr.com/api/v1
STREAMR_WS_URL=wss://www.streamr.com/api/v1/ws
STREAMR_URL=https://www.streamr.com
```

And test that the production canvas:

https://www.streamr.com/canvas/editor/DJ2Ds4ZbRfaXuZ5jC4r7xA-whrguaeTgiaB4tKTqJvxg

works locally:

http://localhost/canvas/editor/DJ2Ds4ZbRfaXuZ5jC4r7xA-whrguaeTgiaB4tKTqJvxg

----

The problem was treating exported, linked variadic ports on a subcanvas as normal linked variadics. Unlike the old UI, linked variadic handling is now "automatic": whenever it sees an input with a `linkedOutput`, it will try to create that output if one does not exist. So when it sees the subcanvas module's variadic input, it can't find the associated output because it exists in the subcanvas modules rather than in the current canvas modules, which then causes an error in the UI because the module isn't configured to handle adding/removing variadic ports. 

The key fix here was to **disable all variadic behaviour when dealing with an embedded subcanvas.** 
323c9b4

In the process of trying to nail down exactly what the problem was, I added a bunch of canvas/variadic tests and thus this PR also includes some minor canvas handling & behaviour fixes:

* Modules can now connect to ports on themselves. We don't currently display this very well but it's not forbidden by the backend implementation so we should allow it.
* Now adds an empty variadic port after exporting last variadic port. This matches old UI behaviour, previously only happened on connection.
* Fixed issue where canvases ended up with a key "undefined" with a value `undefined`
* Fixed `requiresConnection` flag being set on `port.variadic` instead of `port`.
* Improved port longName generation.
* Made linked input/output ports share the same name. Worked before but this matches old UI behaviour.
